### PR TITLE
Add option to exclude individual tests

### PR DESCRIPTION
**Pull request by aalbu**
Passing arguments `-e MyTest.myTestMethod` will cause myTestMethod()
to be excluded from execution (similar to -x for groups)

### CODE REVIEW

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 1]
```diff
@@ -83,33 +92,29 @@ public boolean includeMethod(IMethodSelectorContext context, ITestNGMethod metho
...
                 includeBasedOnGroups(testMetadata) &&
                 !excludeBasedOnGroups(testMetadata);
     }
 
     private boolean includeBasedOnTestName(TestMetadata testMetadata)
     {
-        if (!testNamesToRun.isPresent()) {
-            return true;
-        }
+        return testNamesToRun.map(strings -> indexOf(strings, testMetadata.testName::contains) != -1).orElse(true);
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java:107]: Extract refactoring to a separate commit

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 2]
```diff
@@ -65,14 +67,21 @@
...
+                getListSystemProperty(TEST_NAMES_TO_EXCLUDE_PROPERTY),
                 getListSystemProperty(TEST_GROUPS_TO_RUN_PROPERTY),
                 getListSystemProperty(TEST_GROUPS_TO_EXCLUDE_PROPERTY),
                 new TestMetadataReader());
     }
 
-    public TestNameGroupNameMethodSelector(Optional<Set<String>> testNamesToRun, Optional<Set<String>> testGroupsToRun, Optional<Set<String>> testGroupsToExclude, TestMetadataReader testMetadataReader)
+    public TestNameGroupNameMethodSelector(
+            Optional<Set<String>> testNamesToRun,
+            Optional<Set<String>> testNamesToExclude,
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: Is there a semantics difference between `Optional.empty()` and `Optional.of(Set.of())` for `testNamesToExclude`?

Seems none to me.

Same for `testGroupsToExclude`, so you can have a prep commit removing unnecessary Optional usage and then another one adding `testNamesToExclude`

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 3]
```diff
@@ -83,33 +92,29 @@ public boolean includeMethod(IMethodSelectorContext context, ITestNGMethod metho
...
-        if (!testNamesToRun.isPresent()) {
-            return true;
-        }
+        return testNamesToRun.map(strings -> indexOf(strings, testMetadata.testName::contains) != -1).orElse(true);
+    }
 
-        return indexOf(testNamesToRun.get(), testName -> testMetadata.testName.contains(testName)) != -1;
+    private boolean excludeBasedOnName(TestMetadata testMetadata)
+    {
+        return testNamesToExclude.filter(strings -> indexOf(strings, testMetadata.testName::contains) != -1).isPresent();
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: substring matching is dangerous here.

eg i want to exclude `MyClass.testFoo` but i don't want to exlucde `MyClass.testFooBar`

also, we could require that the exclude comes qualified with simple class name (`MyClass.testFoo` instead of `testFoo`).

if `testMetadata.testName` is not fully qualified, then equals would be a good fit here.
otherwise
```
equals || endswith("." + exclusion)
```

File: tempto-examples/src/main/java/io/prestosql/tempto/examples/TemptoExamples.java [hunk 4]
```diff
@@ -30,10 +33,15 @@ public static void main(String[] args)
                 .setTestsPackage("io.prestosql.tempto.examples,io.prestosql.tempto.another.examples", false)
                 .setConfigFile(DEFAULT_TEST_CONFIGURATION_LOCATION, true)
                 .build();
-        TemptoRunner.runTempto(parser, args);
+        TemptoRunner.runTempto(parser, concat(args, "-e", "ExclusionTest.failingTest"));
```
Comment (findepi) [inline tempto-examples/src/main/java/io/prestosql/tempto/examples/TemptoExamples.java]: There is probably a script which invokes here, add -e there

**Review by findepi — APPROVED** [review]
nice!

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java (hunk 3 above)
Comment (aalbu) [reply tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: I was trying to make this symmetric to `-t` (test inclusion is also substring based), just like `-g` and `-x` are.  But I'll change it to match method names exactly.

**Review by kokosing — APPROVED** [review]
Can you please add example usage to: https://github.com/prestosql/tempto/blob/master/.travis.yml#L15

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 5]
```diff
@@ -54,9 +55,9 @@
     public static final String TEST_GROUPS_TO_RUN_PROPERTY = "io.prestosql.tempto.groups";
     public static final String TEST_GROUPS_TO_EXCLUDE_PROPERTY = "io.prestosql.tempto.exclude_groups";
 
-    private final Optional<Set<String>> testNamesToRun;
-    private final Optional<Set<String>> testGroupsToRun;
-    private final Optional<Set<String>> testGroupsToExclude;
+    private final Set<String> testNamesToRun;
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: This is a semantics change for `testNamesToRun`.
-- previously, Optional.empty() meant: no limit and Optional.of(Set.of()) meant: run nothing.

In my comment i meant `testGroupsToExclude` only.

Also, i think `testNamesToRun = Optional.empty()` (the old way) is a better way of saying "not specified" than `testNamesToRun = Set.of()` (the new way)

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 6]
```diff
@@ -97,6 +109,15 @@ private boolean includeBasedOnTestName(TestMetadata testMetadata)
         return indexOf(testNamesToRun, testMetadata.testName::contains) != -1;
     }
 
+    private boolean excludeBasedOnName(TestMetadata testMetadata)
+    {
+        if (testNamesToExclude.isEmpty()) {
+            return false;
+        }
+
+        return indexOf(testNamesToExclude, testMetadata.testName::equals) != -1;
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: `testNamesToExclude.contains(testMetadata.testName)` ?

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java (hunk 5 above)
Comment (aalbu) [reply tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: Oh, you are right.  I didn't read that carefully.

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 7]
```diff
@@ -56,26 +58,30 @@
 
     private final Optional<Set<String>> testNamesToRun;
     private final Optional<Set<String>> testGroupsToRun;
-    private final Optional<Set<String>> testGroupsToExclude;
+    private final Set<String> testGroupsToExclude;
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java:66]: > Remove unnecessary optionals for representing tests to be included/excluded

update the comjmit message

File: tempto-core/src/test/groovy/io/prestosql/tempto/runner/TestNameGroupNameMethodSelectorTest.groovy [hunk 8]
```diff
@@ -67,4 +68,14 @@ class TestNameGroupNameMethodSelectorTest
...
+
+    private static Set<String> asSet(List<String> strings)
+    {
+        if (strings == null) {
+            return []
+        }
+        else {
+            return strings as Set
+        }
+    }
```
Comment (findepi) [inline tempto-core/src/test/groovy/io/prestosql/tempto/runner/TestNameGroupNameMethodSelectorTest.groovy:87]: ImmutableSet.copyOf(strings) 

(if we have guava on classpath)

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 9]
```diff
@@ -89,35 +95,36 @@ public boolean includeMethod(IMethodSelectorContext context, ITestNGMethod metho
...
-            return true;
-        }
-        return !intersection(testMetadata.testGroups, testGroupsToRun.get()).isEmpty();
+        return testGroupsToRun.map(strings -> !intersection(testMetadata.testGroups, strings).isEmpty()).orElse(true);
     }
 
     private boolean excludeBasedOnGroups(TestMetadata testMetadata)
     {
-        if (!testGroupsToExclude.isPresent()) {
+        if (testGroupsToExclude.isEmpty()) {
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: You can remove the if now

File: tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java [hunk 10]
```diff
@@ -98,6 +107,15 @@ private boolean includeBasedOnTestName(TestMetadata testMetadata)
         return testNamesToRun.map(strings -> indexOf(strings, testMetadata.testName::contains) != -1).orElse(true);
     }
 
+    private boolean excludeBasedOnName(TestMetadata testMetadata)
+    {
+        if (testNamesToExclude.isEmpty()) {
+            return false;
+        }
```
Comment (findepi) [inline tempto-core/src/main/java/io/prestosql/tempto/internal/listeners/TestNameGroupNameMethodSelector.java]: redundant

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
Merged, thanks!

**Comment by findepi** [comment]
Released as 176
